// @node-red/editor-api/lib/index.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

 /**
  * This module provides an Express application to serve the Node-RED editor.
  *
  * It implements the Node-RED HTTP Admin API the Editor uses to interact
  * with the Node-RED runtime.
  *
  * @namespace @node-red/editor-api
  */

var bodyParser = require("body-parser");
var passport = require('passport');
var fs = require('fs');
var path = require('path');

var auth = require("./auth");
var apiUtil = require("./util");

var adminApp;
var server;
var editor;

/**
 * Initialise the module.
 * @param  {Object}     settings   The runtime settings
 * @param  {HTTPServer} _server     An instance of HTTP Server
 * @param  {Storage}    storage    An instance of Node-RED Storage
 * @param  {Runtime}    runtimeAPI An instance of Node-RED Runtime
 * @memberof @node-red/editor-api
 */
function init(settings,_server,storage,runtimeAPI) {
    server = _server;
    if (settings.httpAdminRoot !== false) {
        adminApp = apiUtil.createExpressApp(settings);

        var cors = require('cors');
        var corsHandler = cors({
           origin: "*",
           methods: "GET,PUT,POST,DELETE"
        });
        adminApp.use(corsHandler);

        if (settings.httpAdminMiddleware) {
            if (typeof settings.httpAdminMiddleware === "function" || Array.isArray(settings.httpAdminMiddleware)) {
                adminApp.use(settings.httpAdminMiddleware);
            }
        }

        auth.init(settings,storage);

        var maxApiRequestSize = settings.apiMaxLength || '5mb';
        adminApp.use(bodyParser.json({limit:maxApiRequestSize}));
        adminApp.use(bodyParser.urlencoded({limit:maxApiRequestSize,extended:true}));

        // ====== START TRIPLETS API INTEGRATION ======
        // Register the triplets API endpoint
        adminApp.get("/api/load_schema", function(req, res) {
            console.log("Schema API endpoint called");
            
            try {
                // Define the path to the data file
                var dataFilePath = path.join(settings.userDir || process.cwd(), 'public', 'data', 'schema.json');
                console.log("Looking for schema data at:", dataFilePath);
                
                // Check if the file exists
                if (!fs.existsSync(dataFilePath)) {
                    console.log("schema file not found");
                    return res.status(404).json({
                        success: false,
                        error: "Data file not found"
                    });
                }
                
                var dataFileContents = fs.readFileSync(dataFilePath, 'utf8');
                
                var schemaData = JSON.parse(dataFileContents);
                console.log("schemaData", typeof(schemaData));
                
                var response = {
                    success: true,
                    timestamp: new Date().toISOString(),
                    ...schemaData,
                    count: (schemaData.entities ? schemaData.entities.length : 0) + 
                           (schemaData.relations ? schemaData.relations.length : 0)
                };
                
                
                          
                res.json(response);
                
            } catch(err) {
                // Handle any errors
                console.error("Error in /api/load_schema endpoint:", err);
                res.status(500).json({
                    success: false,
                    error: err.toString()
                });
            }
        });

        // Add MQTT Configuration endpoints
        adminApp.get("/settings/global-mqtt", function(req, res) {
            try {
                // Get MQTT config from settings
                const config = settings.globalMqttConfig || {
                    broker: "localhost",
                    port: "1883"
                };
                console.log("Getting MQTT config:", config);
                res.json(config);
            } catch (err) {
                console.error("Error getting MQTT config:", err);
                res.status(500).json({ error: "Internal server error", message: err.message });
            }
        });

        adminApp.post("/settings/global-mqtt", function(req, res) {
            try {
                console.log("Received request to save MQTT config");
                console.log("Request body:", req.body);
                
                if (!req.body || !req.body.broker || !req.body.port) {
                    throw new Error("Invalid MQTT configuration. Both broker and port are required.");
                }
                
                const config = {
                    broker: req.body.broker,
                    port: req.body.port
                };

                // Save to runtime settings
                settings.globalMqttConfig = config;

                // Emit event for MQTT config update
                runtimeAPI.events.emit("global-mqtt-config-update", config);
                
                console.log("MQTT config saved successfully:", config);
                res.json({ status: "success", config: config });
            } catch (err) {
                console.error("Error saving MQTT config:", err);
                res.status(400).json({ error: "Bad request", message: err.message });
            }
        });

        adminApp.get("/api/load_doxigen", function(req, res) {
            console.log("Doxigen API endpoint called");
            
            try {
                // Define the path to the data file
                var dataFilePath = path.join(settings.userDir || process.cwd(), 'public', 'data', 'doxigen_data.json');
                console.log("Looking for doxigen data at:", dataFilePath);
                
                // Check if the file exists
                if (!fs.existsSync(dataFilePath)) {
                    console.log("Doxigen data file not found");
                    return res.status(404).json({
                        success: false,
                        error: "Data file not found"
                    });
                }
                
                var dataFileContents = fs.readFileSync(dataFilePath, 'utf8');
                
                var doxigenData = JSON.parse(dataFileContents);
                
                var response = {
                    success: true,
                    timestamp: new Date().toISOString(),
                    ...doxigenData,
                    count: (doxigenData.entities ? doxigenData.entities.length : 0) + 
                           (doxigenData.relations ? doxigenData.relations.length : 0)
                };
                
                
                
                res.json(response);
                
            } catch(err) {
                // Handle any errors
                console.error("Error in /api/load_doxigen endpoint:", err);
                res.status(500).json({
                    success: false,
                    error: err.toString()
                });
            }
        });
        
        adminApp.get("/auth/login",auth.login,apiUtil.errorHandler);
        if (settings.adminAuth) {
            if (settings.adminAuth.type === "strategy") {
                auth.genericStrategy(adminApp,settings.adminAuth.strategy);
            } else if (settings.adminAuth.type === "credentials") {
                adminApp.use(passport.initialize());
                adminApp.post("/auth/token",
                    auth.ensureClientSecret,
                    auth.authenticateClient,
                    auth.getToken,
                    auth.errorHandler
                );
            } else if (settings.adminAuth.tokens) {
                adminApp.use(passport.initialize());
            }
            adminApp.post("/auth/revoke",auth.needsPermission(""),auth.revoke,apiUtil.errorHandler);
        }

        // Editor
        if (!settings.disableEditor) {
            editor = require("./editor");
            var editorApp = editor.init(server, settings, runtimeAPI);
            adminApp.use(editorApp);
        }

        if (settings.httpAdminCors) {
            var corsHandler = cors(settings.httpAdminCors);
            adminApp.use(corsHandler);
        }

        var adminApiApp = require("./admin").init(settings, runtimeAPI);
        adminApp.use(adminApiApp);
    } else {
        adminApp = null;
    }
}

/**
 * Start the module.
 * @return {Promise} resolves when the application is ready to handle requests
 * @memberof @node-red/editor-api
 */
async function start() {
    if (editor) {
        return editor.start();
    }
}

/**
 * Stop the module.
 * @return {Promise} resolves when the application is stopped
 * @memberof @node-red/editor-api
 */
async function stop() {
    if (editor) {
        editor.stop();
    }
}

module.exports = {
    init,
    start,
    stop,

    /**
    * @memberof @node-red/editor-api
    * @mixes @node-red/editor-api_auth
    */
    auth: {
        needsPermission: auth.needsPermission
    },
    /**
     * The Express app used to serve the Node-RED Editor
     * @type ExpressApplication
     * @memberof @node-red/editor-api
     */
    get httpAdmin() { return adminApp; }
};
